bpf/lib/wireguard.h

```diff
@@ -0,0 +1,78 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Cilium */
+
+#ifndef __WIREGUARD_H_
+#define __WIREGUARD_H_
+
+#include <bpf/ctx/ctx.h>
+#include <bpf/api.h>
+
+#include "tailcall.h"
+#include "common.h"
+#include "overloadable.h"
+
+static __always_inline int
+wg_maybe_redirect_to_encrypt(struct __ctx_buff *ctx)
+{
+	struct remote_endpoint_info *dst;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data, *data_end;
+	__u16 proto = 0;
+#ifdef ENABLE_IPV6
+	struct ipv6hdr *ip6;
+#endif
+#ifdef ENABLE_IPV4
+	struct iphdr *ip4;
+#endif
+
+	if (!validate_ethertype(ctx, &proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+#ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		if (!revalidate_data(ctx, &data, &data_end, &ip6))
+			return DROP_INVALID;
+		dst = lookup_ip6_remote_endpoint((union v6addr *)&ip6->daddr);
+#ifndef ENABLE_NODE_ENCRYPTION
+		src = lookup_ip6_remote_endpoint((union v6addr *)&ip6->saddr);
+#endif /* ENABLE_NODE_ENCRYPTION */
+		break;
+#endif
+#ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+		dst = lookup_ip4_remote_endpoint(ip4->daddr);
+#ifndef ENABLE_NODE_ENCRYPTION
+		src = lookup_ip4_remote_endpoint(ip4->saddr);
+#endif /* ENABLE_NODE_ENCRYPTION */
+		break;
+#endif
+	default:
+		goto out;
+	}
+
+	/* Redirect to the WireGuard tunnel device if the encryption is
+	 * required.
+	 *
+	 * After the packet has been encrypted, the WG tunnel device
+	 * will set the MARK_MAGIC_WG_ENCRYPTED skb mark. So, to avoid
```

Comment:
I'm confused how this gets set, is it purely in the wg implementation or do we have bpf on wg interface?

Reply:
That actually happens in a later commit. Grep for `FirewallMark`. But yeah, this character being used in the story before it's introduced leaves the reader somewhat confused :sweat_smile: